### DISCUSSION TITLE
Knowing if it is Warmup time?

### DISCUSSION THREAD

**Discussion by antero-lukkonen** [roflmuffin/CounterStrikeSharp#674] [Q&A] [answered]
Hi, my usecase is to know for statistics purposes, if a kill happened during warmup.

For that i have tried to access the gamerules entity, since it has helpful fields, at least by name.

I find that the gamerules entity fields are mostly uninitialized. 

I have tried to access them in the onroundstart, onplayerdeath, onplayerhurt events.

Is the gamerules entity supposed to be used for that?

My next attempt will be to keep state myself based on various events.. and i hesitate since i fear its not going to be trivial ( edge case state transitions i dont know ).

**Comment by KillStr3aK** [comment]
Hey!

yes, gamerules can be used to retrieve various game phases, mind sharing the code you've tried? it might go wrong somewhere

also what gametype and gamemode?

**Reply by antero-lukkonen** [reply]
Thank you for your reply.
cssharp 287

game_type is: 0
game_mode is: 1

I am attempting something very simple:

```
        // Handler to test the gamerules entity usage.
        this.RegisterEventHandler<EventRoundStart>((gameEvent, _) =>
            {
                var gameRules = Utilities.FindAllEntitiesByDesignerName<CBaseEntity>("cs_gamerules").Single().As<CCSGameRules>();
           
                Console.WriteLine("Expecting iswarmup to be true when warmup message is displayed - " + gameRules.WarmupPeriod);
                Console.WriteLine("Expecting hasmatchstarted to be true - " + gameRules.HasMatchStarted);
                Console.WriteLine("MaxNumCTs - " + gameRules.MaxNumCTs);
                Console.WriteLine("Game phase - " + gameRules.GamePhase);
                return HookResult.Continue;
            }, HookMode.Post);
           
           ```  
Console output for these lines looks as follows for the Warmup period as well as when the "real" round starts:

2024-11-16 14:49:07 EventRoundStart:{"WarmupPeriod":false,"HasMatchStarted":false,"MatchStartTime":0,"MaxNumCTs":1952542070}
2024-11-16 14:49:07 Expecting iswarmup to be true when warmup message is displayed - False
2024-11-16 14:49:07 Expecting hasmatchstarted to be true - False
2024-11-16 14:49:07 MaxNumCTs - 1952542070
2024-11-16 14:49:07 Game phase - 0

I have a similar handler attached to EventPlayerHurt and the output is the same.

**Comment by antero-lukkonen** [comment, marked as answer]
After experimenting and stumbling on these docs: https://cs2.poggu.me/dumped-data/entity-list/
I managed to get gamerules working by changing
`var gameRules = Utilities.FindAllEntitiesByDesignerName<CBaseEntity>("cs_gamerules").Single().As<CCSGameRules>();`
to
`var gameRules = Utilities.FindAllEntitiesByDesignerName<CBaseEntity>("cs_gamerules").Single().As<CCSGameRulesProxy>().GameRules!;`
